packages/financial-templates-lib/src/clients/FinancialContractClient.js

```diff
@@ -159,9 +165,35 @@ class FinancialContractClient {
     ]);
 
     if (this.contractType === "Perpetual") {
-      this.latestCumulativeFundingRateMultiplier = this.toBN(
-        (await this.financialContract.methods.fundingRate().call()).cumulativeMultiplier.rawValue
-      );
+      if (!this.multicallContractAddress) {
+        // If no multicontract address set, then read the contract's on-chain CFRM. Note
+        // that this will not take into account any pending fundingrates that will be
+        // published on the next contract interaction.
+        this.latestCumulativeFundingRateMultiplier = this.toBN(
+          (await this.financialContract.methods.fundingRate().call()).cumulativeMultiplier.rawValue
+        );
+      } else {
+        // Simulate calling `applyFundingRate()` on the perpetual before reading
+        // `fundingRate()`, in order to account for any unpublished, pending funding rates.
+        const applyFundingRateCall = {
+          target: this.financialContract.options.address,
+          callData: this.financialContract.methods.applyFundingRate().encodeABI()
+        };
+        const fundingRateCall = {
+          target: this.financialContract.options.address,
+          callData: this.financialContract.methods.fundingRate().encodeABI()
+        };
+        const outputs = await aggregateTransactionsAndCall(this.multicallContractAddress, this.web3, [
+          applyFundingRateCall,
+          fundingRateCall
+        ]);
+        const fundingRateData = this.web3.eth.abi.decodeParameters(
+          contractMethodABIs["fundingRate"],
+          // The return data for `fundingRate()` is the second return argument.
+          outputs.returnData[1]
```

Comment:
does the returned data not also offer key-values on the returned types? is it _just_ an array of elements? often I've seen these kinds of methods return both the array and a key-valued type. just mentioning as it will be way easier to read if it is not referencing an array index.

Reply:
Here's an example return output:

```
Result {
  '0': '96',
  '1': [
    '0x',
    '0x0000000000000000000000000000000000000000000000000000000000000000756e646566696e65645f46554e44494e475f4944454e544946494552000000000000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000000000000000000000000000000000003adb1765300000000000000000000000000000000000000000000000000000003adb176540000000000000000000000000000000000000000000000000000000000000000'
  ],
  blockNumber: '96',
  returnData: [
    '0x',
    '0x0000000000000000000000000000000000000000000000000000000000000000756e646566696e65645f46554e44494e475f4944454e544946494552000000000000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000000000000000000000000000000000003adb1765300000000000000000000000000000000000000000000000000000003adb176540000000000000000000000000000000000000000000000000000000000000000'
  ]
}
```

You'll notice that the returnData is just a simple array